File: packages/backend/src/app/routers/programs.py
"""Programs CRUD API."""

from fastapi import APIRouter, Depends

from app.deps import get_tenant_id
from app.models.programs import ProgramCreate, ProgramUpdate, ProgramResponse, ProgramListResponse
from app.services import programs as svc
from app.exceptions import NotFoundError

router = APIRouter(prefix="/programs", tags=["Programs"])


@router.get("", response_model=ProgramListResponse)
def list_programs(tenant_id: str = Depends(get_tenant_id)):
    items = svc.list_programs(tenant_id)
    return ProgramListResponse(programs=[ProgramResponse(**x) for x in items])


@router.get("/{program_id}", response_model=ProgramResponse)
def get_program(program_id: str, tenant_id: str = Depends(get_tenant_id)):
    item = svc.get_program(tenant_id, program_id)
    if not item:
        raise NotFoundError("Program not found")
    return ProgramResponse(**item)


@router.post("", response_model=dict, status_code=201)
def create_program(body: ProgramCreate, tenant_id: str = Depends(get_tenant_id)):
    return svc.create_program(tenant_id, body.name or "My Program", body.currency or "INR")


@router.put("/{program_id}", response_model=dict)
def update_program(program_id: str, body: ProgramUpdate, tenant_id: str = Depends(get_tenant_id)):
    return svc.update_program(
        tenant_id,
        program_id,
        name=body.name,
        currency=body.currency,
        earn_rules=body.earnRules,
        burn_rules=body.burnRules,
        tier_config=body.tierConfig,
    )

File: packages/web/src/api/programs.ts
import { apiGet, apiPost, apiPut } from './client'

export interface Program {
  programId: string
  name: string
  currency: string
  earnRules?: Record<string, unknown>
  burnRules?: Record<string, unknown>
  tierConfig?: Record<string, unknown>
  createdAt?: string
  updatedAt?: string
}

export interface ListProgramsResponse {
  programs: Program[]
}

export function listPrograms(tenantId: string, idToken?: string | null): Promise<ListProgramsResponse> {
  return apiGet<ListProgramsResponse>('/api/v1/programs', tenantId, idToken)
}

export function getProgram(tenantId: string, programId: string, idToken?: string | null): Promise<Program> {
  return apiGet<Program>(`/api/v1/programs/${programId}`, tenantId, idToken)
}

export function createProgram(
  tenantId: string,
  body: { name?: string; currency?: string },
  idToken?: string | null
): Promise<{ programId: string; name: string; currency: string }> {
  return apiPost(`/api/v1/programs`, tenantId, body, idToken)
}

export function updateProgram(
  tenantId: string,
  programId: string,
  body: { name?: string; currency?: string; earnRules?: object; burnRules?: object; tierConfig?: object },
  idToken?: string | null
): Promise<{ programId: string; updatedAt: string }> {
  return apiPut(`/api/v1/programs/${programId}`, tenantId, body, idToken)
}
